666: Add CLDR support

toolstack opened the pull request:
Resolves #666.

toolstack committed:
Add CLDR plurals to the locales, formats and projects.

toolstack committed:
Limit export dropdown to only formats that match the plurals type.

toolstack committed:
Add plural support to the Android format.

toolstack committed:
Add a script to generate the locale details.

toolstack committed:
Add examples and number generator for cldr plurals.

Also update code differentiate between the types.

Re-order few/many to match CLDR order in the locales.

toolstack committed:
Update the CLDR plurals script.

toolstack committed:
Update the locale.php file based on the output from the new CLDR plurals script.

toolstack committed:
Make sure the CLDR number list honours the $how_many parameter.

toolstack committed:
Fix the android export test case.

And add a minor comment to the locale generator.

toolstack committed:
Add docblocks and comments to locales.php.

toolstack commented:
A couple of notes on the current state of this:
- I have only added CLDR plurals support to the android format, the Apple strings format remains to be done.
- The test cases haven't been updated to do importing of CLDR plurals yet and the export test case is a bit of a fudge ;)
- ~~There's probably a few more docblocks to be written.~~

toolstack committed:
Coding standard updates.

yoavf commented on bin/locale-generator/locale-generator.php:
```diff
@@ -0,0 +1,139 @@
+/*
+ * Get the CLDR Data.
+ *
+ * By defualt it is read from the plurals.json file in the current directory, you can download
```
typo `defualt`

yoavf commented on bin/locale-generator/locale-generator.php:
```diff
@@ -0,0 +1,139 @@
+// Create a working locales object.
+$locales = new GP_Locales;
+
+// Run through the locales and see if we can find a mathcing CLDR locale.
```
Typo in `mathcing`

yoavf commented on tests/phpunit/testcases/tests_formats/test_format_android.php:
```diff
@@ -24,11 +27,22 @@ function test_export() {
 
 		foreach( $this->entries as $sample ) {
-			list( $context, $original, $translation ) = $sample;
+			list( $context, $original, $translation, $comment ) = $sample;
```
Is the addition of the comment intentional in this patch?

toolstack replied:
Yes, but only because it fixes an existing bug in the test case.

yoavf reviewed (commented):
Really happy to see this PR. Left a few comments on a few very minor issues spotted while taking a quick look

toolstack committed:
Add plurals type to ngx format.

toolstack committed:
Add downloading of the CLDR plurals file if a local copy does not exist.

toolstack committed:
Add columns to the locales page to indicate plurals support.

toolstack committed:
Coding standard updates.

ocean90 commented on gp-includes/template.php:
```diff
@@ -458,6 +458,26 @@ function gp_projects_dropdown( $name_and_id, $selected_project_id = null, $attrs
+ *
+ * @return string HTML markup for a select element.
+ */
+function gp_plurals_dropdown( $name_and_id, $selected_plural_type = null, $attrs = array() ) {
```
Let's name this `gp_plural_types_dropdown()`.

toolstack replied:
:+1:

ocean90 commented on gp-templates/project-form.php:
```diff
@@ -12,6 +12,9 @@
 	<dd><textarea name="project[description]" rows="4" cols="40" id="project[description]"><?php echo esc_html( $project->description ); ?></textarea></dd>
 
+	<dt><label for="project[plurals_type]"><?php _e( 'Plurals Type', 'glotpress' ); ?></label></dt>
+	<dd><?php echo gp_plurals_dropdown( 'project[plurals_type]', $project->plurals_type, array() ); // WPCS: XSS ok. ?></dd>
```
`gp_plurals_dropdown()` should be added to `customAutoEscapedFunctions` in phpcs.xml.

toolstack replied:
:+1:

ocean90 commented on bin/locale-generator/locale-generator.php:
```diff
@@ -0,0 +1,146 @@
+		} // End foreach().
+
+		// Add some space between locales for easier reading.
+		echo "\n"; // WPCS: XSS ok.
```
Can we ignore the variants part in this PR for now?

toolstack replied:
Not easily without breaking one or both of the patches :)

They both do a lot of work on the locales file so separating them is hard.

ocean90 commented on bin/locale-generator/locale-generator.php:
```diff
@@ -0,0 +1,146 @@
+					}
+				}
+			}
+		} // End foreach().
```
These are no longer required and can be removed.

toolstack replied:
:+1:

ocean90 commented on bin/locale-generator/locale-generator.php:
```diff
@@ -0,0 +1,146 @@
+		$root_var_name = str_replace( '-', '_', $locale->variant_root );
+
+		// Output the first line to 'create' the GP_Locale object for this locale.
+		echo "\t\t\${$var_name} = new GP_Locale();\n"; // WPCS: XSS ok.
```
Remove all `// WPCS: XSS ok.` comments and add the bin directory as a exclude-pattern to phpcs.xml.

toolstack replied:
:+1:

ocean90 commented on bin/locale-generator/locale-generator.php:
```diff
@@ -0,0 +1,146 @@
+					}
+				}
+			}
+		} // End foreach().
```
These are no longer required and can be removed.

toolstack replied:
:+1:

toolstack committed:
Various updates.

Rename dropdown function and PHP CS rule additions.

Fix error.

toolstack committed:
Update phpcs exclusion.

toolstack committed:
Fix inverted logic check.

toolstack committed:
Add file format definition link to comments.

toolstack committed:
First cut at stringdict support.

Import not yet implemented.

toolstack committed:
Coding standard updates.

ocean90 commented on gp-templates/project-form.php:
```diff
@@ -12,6 +12,9 @@
 	<dd><textarea name="project[description]" rows="4" cols="40" id="project[description]"><?php echo esc_html( $project->description ); ?></textarea></dd>
 
+	<dt><label for="project[plurals_type]"><?php _e( 'Plurals Type', 'glotpress' ); ?></label></dt>
+	<dd><?php echo gp_plural_types_dropdown( 'project[plurals_type]', $project->plurals_type, array() ); // WPCS: XSS ok. ?></dd>
```
The `// WPCS: XSS ok` seems to be obsolete.

ocean90 reviewed (changes requested):
The current locale alignment doesn't match our code styling. Example:

```
<?php
$av = new GP_Locale();
$av->english_name                         = 'Avaric';
$av->native_name                          = 'авар мацӀ';
$av->text_direction                       = 'ltr';
$av->lang_code_iso_639_1                  = 'av';
$av->lang_code_iso_639_2                  = 'ava';
$av->slug                                 = 'av';
$av->nplurals                             = '2';
$av->plural_expression                    = 'n != 1';
```

produces:

```
-----------------------------------------------------------------------------------------------------------------
FOUND 0 ERRORS AND 9 WARNINGS AFFECTING 9 LINES
-----------------------------------------------------------------------------------------------------------------
  2 | WARNING | [x] Equals sign not aligned with surrounding assignments; expected 22 spaces but found 1 space
  3 | WARNING | [x] Equals sign not aligned with surrounding assignments; expected 8 spaces but found 25 spaces
  4 | WARNING | [x] Equals sign not aligned with surrounding assignments; expected 9 spaces but found 26 spaces
  5 | WARNING | [x] Equals sign not aligned with surrounding assignments; expected 6 spaces but found 23 spaces
  6 | WARNING | [x] Equals sign not aligned with surrounding assignments; expected 1 space but found 18 spaces
  7 | WARNING | [x] Equals sign not aligned with surrounding assignments; expected 1 space but found 18 spaces
  8 | WARNING | [x] Equals sign not aligned with surrounding assignments; expected 16 spaces but found 33 spaces
  9 | WARNING | [x] Equals sign not aligned with surrounding assignments; expected 12 spaces but found 29 spaces
 10 | WARNING | [x] Equals sign not aligned with surrounding assignments; expected 3 spaces but found 20 spaces
-----------------------------------------------------------------------------------------------------------------
PHPCBF CAN FIX THE 9 MARKED SNIFF VIOLATIONS AUTOMATICALLY
-----------------------------------------------------------------------------------------------------------------
```

It needs to be:

```
$av                      = new GP_Locale();
$av->english_name        = 'Avaric';
$av->native_name         = 'авар мацӀ';
$av->text_direction      = 'ltr';
$av->lang_code_iso_639_1 = 'av';
$av->lang_code_iso_639_2 = 'ava';
$av->slug                = 'av';
$av->nplurals            = '2';
$av->plural_expression   = 'n != 1';
```

ocean90 commented on locales/locales.php:
```diff
@@ -87,454 +433,772 @@ public function index_for_number( $number ) {
+		$af->cldr_code                            = 'af';
+		$af->cldr_nplurals                        = '2';
+		$af->cldr_plural_expressions['one']       = 'n = 1 @integer 1 @decimal 1.0, 1.00, 1.000, 1.0000';
+		$af->cldr_plural_expressions['other']     = ' @integer 0, 2~16, 100, 1000, 10000, 100000, 1000000, … @decimal 0.0~0.9, 1.1~1.6, 10.0, 100.0, 1000.0, 10000.0, 100000.0, 1000000.0, …';
```
Is the leading whitespace required?

toolstack replied:
I don't think so but it's in the original JSON file from the CLDR repo so I think it's safer to leave in.

GaryJones commented on gp-includes/formats/format-resx.php:
```diff
@@ -162,6 +274,11 @@ private function add_schema_info() {
 	}
 
+	/**
+	 * Cteate the .resx schema declaration.  .resx file have a standard scheam declaration.
```
Several typos here.

toolstack replied:
Fixed.

GaryJones commented on gp-includes/formats/format-resx.php:
```diff
@@ -79,27 +154,64 @@ public function read_originals_from_file( $file_name ) {
 
-
+	/**
+	 * Cteate the .resx schema info.  .resx file have a standard scheam block.
```
Several typos here.

toolstack replied:
Fixed.

GaryJones commented on phpcs.xml:
```diff
@@ -39,9 +39,32 @@
+		in the current version of phpcs (see https://github.com/squizlabs/PHP_CodeSniffer/issues/1731)
+		which breaks this when a single file name is passed to be checked (for example with TravisCI).
+
+		A fix has been committed and will be included in phpcs version 3.2, however this is no release
```
PHPCS 3.2 was released in December 2017, and the latest release is 3.2.3, with 3.3.0 not too far away. It would be reasonable to make `^3.2` the minimum version of PHPCS required for devs to contribute to GlotPress.

toolstack replied:
This should have automatically taken care of itself, we just have to remove the extra lines now as they were only needed before 3.2 was released.

toolstack committed:
Add padding to locale object creation lines.

toolstack committed:
Update locale file.

toolstack committed:
Fixed typo's in comment.

toolstack committed:
Regerenate locales file.

toolstack committed:
Update locale generation script

Now appends the code around the locales list to make life easier.

toolstack committed:
Coding standard updates.

toolstack committed:
Bump DB version after notes merge.

toolstack committed:
Update android test cases.

toolstack committed:
Update locale generator.

To create no-variants version of the locales file and also write both out to files instead of the console.  Also make a backup just in case before deleting the old file.

toolstack committed:
Coding standard updates.

amieiro commented:
This branch has a lot of conflicts, so we have decided in the triage session we are going to close it.
